builder/qemu/step_port_forward.go

```diff
@@ -0,0 +1,60 @@
+package qemu
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/packer/common/net"
+	"github.com/hashicorp/packer/helper/multistep"
+	"github.com/hashicorp/packer/packer"
+)
+
+// This step adds a NAT port forwarding definition so that SSH or WinRM is available
+// on the guest machine.
+type stepPortForward struct {
+	l *net.Listener
+}
+
+func (s *stepPortForward) Run(ctx context.Context, state multistep.StateBag) multistep.StepAction {
+	config := state.Get("config").(*Config)
+	ui := state.Get("ui").(packer.Ui)
+
+	guestPort := config.CommConfig.Comm.Port()
+	commHostPort := guestPort
+	if !config.CommConfig.SkipNatMapping {
+		log.Printf("Looking for available communicator (SSH, WinRM, etc) port between %d and %d", config.CommConfig.HostPortMin, config.CommConfig.HostPortMax)
+		var err error
+		s.l, err = net.ListenRangeConfig{
+			Addr:    config.VNCBindAddress,
+			Min:     config.CommConfig.HostPortMin,
+			Max:     config.CommConfig.HostPortMax,
+			Network: "tcp",
+		}.Listen(ctx)
+		if err != nil {
+			err := fmt.Errorf("Error finding port: %s", err)
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+		s.l.Listener.Close() // free port, but don't unlock lock file
+		commHostPort = s.l.Port
+		ui.Say(fmt.Sprintf("Found port for communicator (SSH, WinRM, etc): %d.", commHostPort))
+
+	} else {
```

Comment:
This could be a guard clause that leaves the step early, which would probably be easier to read than this big if/else.